### PR TITLE
Fix modelpart io

### PR DESCRIPTION
**Pull request by roigcarlo**
That should fix the problem with #1427 with the new implementation.
I will upload some more tests for that when I find a elegant solution to compare modelparts form c++. thx again @msandre

### CODE REVIEW

File: kratos/sources/model_part_io.cpp [hunk 1]
```diff
@@ -260,14 +250,13 @@ namespace Kratos
         // We are going to procede like the following, we are going to iterate over all the elements and compare with the components, we will save the type and we will compare until we get that the type of element has changed
         if (rThisElements.size() > 0) {
             std::string element_name;
+
+            auto it_element = rThisElements.begin();
             auto elements_components = KratosComponents<Element>::GetComponents();
```
Comment (msandre) [inline kratos/sources/model_part_io.cpp:255]: auto& to avoid copy constructor.

File: kratos/sources/model_part_io.cpp [hunk 2]
```diff
@@ -366,14 +349,13 @@ namespace Kratos
         
         if (rThisConditions.size() > 0) {
             std::string condition_name;
+
+            auto it_condition = rThisConditions.begin();
             auto conditions_components = KratosComponents<Condition>::GetComponents();
```
Comment (msandre) [inline kratos/sources/model_part_io.cpp:344]: same here.

**Review by msandre — DISMISSED** [review]
Changes look good, thx.

**Comment by roigcarlo** [comment]
Ouch, missed those thx

**Review by loumalouomega — DISMISSED** [review]

**Comment by roigcarlo** [comment]
Seems we cannot use `auto &` here but not sure why...
**Edit**: I just removed them. No use to have them in a temp var with the new compare function....

**Comment by msandre** [comment]
I was meaning to avoid the copy for the components which is something like a map of string to pointer. I need the same search function so was thinking we could put it in CompareElementAndConditionUtility::GetRegisteredName(const Element& rElement, std::string& rName); and one for condition.

**Comment by roigcarlo** [comment]
So basically move all the loop inside the utility and hide all the component implementation? (Just to double check, not my best week...)
I think it is a really good idea 👍

**Comment by msandre** [comment]
Yes, basically the part that searches through the components for the corresponding string name. It appears twice in the WriteElements(). I'm using it in an application in my working branch right now until this gets merged. These are the functions:

```
    static void GetRegisteredName(const Element& rElement, std::string& rName)
    {
        KRATOS_TRY;

        const auto& components = KratosComponents<Element>::GetComponents();
        for(auto it = components.begin(); it != components.end(); ++it) 
        {
            if (IsSame(*(it->second), rElement))
            {
                rName = it->first;
                return;
            }
        }

        KRATOS_ERROR << "Element \"" << typeid(rElement).name() << "\" not found!" << std::endl;

        KRATOS_CATCH("");
    }

    static void GetRegisteredName(const Condition& rCondition, std::string& rName)
    {
        KRATOS_TRY;

        const auto& components = KratosComponents<Condition>::GetComponents();
        for(auto it = components.begin(); it != components.end(); ++it) 
        {
            if (IsSame(*(it->second), rCondition))
            {
                rName = it->first;
                return;
            }
        }

        KRATOS_ERROR << "Condition \"" << typeid(rCondition).name() << "\" not found!" << std::endl;

        KRATOS_CATCH("");
    }
```

If you agree, I would suggest adding them to the core utility in this pr.

File: kratos/sources/model_part_io.cpp [hunk 3]
```diff
@@ -262,12 +252,9 @@ namespace Kratos
...
             auto elements_components = KratosComponents<Element>::GetComponents();
             
-            auto element_type = rThisElements.begin()->GetGeometry().GetGeometryType();
-            
             // Fisrt we do the first element
             for(auto it_comp = elements_components.begin(); it_comp != elements_components.end() ; it_comp++) {
-                if (IsSameElement<decltype(*(rThisElements.begin())), decltype(*(it_comp->second))>::value &&
-                    (element_type == (it_comp->second)->GetGeometry().GetGeometryType())) {
+                if(CompareElementsAndConditionsUtility::IsSame(*(rThisElements.begin()), *it_comp->second)) {
                     element_name = it_comp->first;
```
Comment (pooyan-dadvand) [inline kratos/sources/model_part_io.cpp]: If you have a utility for comparing elements I would add the geometry comparison to it.

File: kratos/sources/model_part_io.cpp [hunk 4]
```diff
@@ -262,12 +252,9 @@ namespace Kratos
             std::string element_name;
             auto elements_components = KratosComponents<Element>::GetComponents();
             
-            auto element_type = rThisElements.begin()->GetGeometry().GetGeometryType();
-            
             // Fisrt we do the first element
             for(auto it_comp = elements_components.begin(); it_comp != elements_components.end() ; it_comp++) {
-                if (IsSameElement<decltype(*(rThisElements.begin())), decltype(*(it_comp->second))>::value &&
-                    (element_type == (it_comp->second)->GetGeometry().GetGeometryType())) {
+                if(CompareElementsAndConditionsUtility::IsSame(*(rThisElements.begin()), *it_comp->second)) {
```
Comment (pooyan-dadvand) [inline kratos/sources/model_part_io.cpp]: Why `IsSame` is not a method of element?

File: kratos/sources/model_part_io.cpp [hunk 5]
```diff
@@ -16,6 +16,7 @@
 #include "input_output/logger.h"
 #include "utilities/quaternion.h"
 #include "utilities/openmp_utils.h"
+#include "utilities/compare_elements_and_conditions_utility.h"
```
Comment (pooyan-dadvand) [inline kratos/sources/model_part_io.cpp:19]: Have you added this file?

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
I would add the `IsSameType` and  `HasSameGeometryType`  method to the element base class

@KratosMultiphysics/technical-committee

File: kratos/sources/model_part_io.cpp (hunk 5 above)
Comment (roigcarlo) [reply kratos/sources/model_part_io.cpp:19]: This was added here #1434

File: kratos/sources/model_part_io.cpp (hunk 3 above)
Comment (roigcarlo) [reply kratos/sources/model_part_io.cpp]: It already does

**Comment by roigcarlo** [comment]
@pooyan-dadvand It was added here #1434. It is in a different file because it compares both elements and conditions. Also I think @msandre idea is to add there several other utilities there. For the record I like the idea of having the `GetRegisteredName`

**Comment by pooyan-dadvand** [comment]
I found the #1434 after my message. Still I think that these methods are for element and conditions.

**Comment by msandre** [comment]
I agree with the point of @pooyan-dadvand. Also the function names are more descriptive than what I had.

**Comment by pooyan-dadvand** [comment]
After discussion in TC, the conclusion is to add `IsSameType` and  `HasSameGeometryType` to `GeometricalObject` and leave the `GetRegisteredName` in the utility to be moved to future `Registry` class

**Comment by roigcarlo** [comment]
@msandre I will move the functions after lunch, do you want me to add the GetRegistered name as well?

**Comment by msandre** [comment]
It makes sense to me to include it in this pr since it can be used with the IsSameType() and HasSameGeometryType() to refactor WriteElements() and WriteConditions(). Otherwise I could add it in a separate pr.

**Comment by roigcarlo** [comment]
Nono, here its ok, I will add myself all together :+1:

**Review by pooyan-dadvand — COMMENTED** [review]
Do we need all the versions taking pointer and reference?
I would put only the one taking references. @msandre and @roigcarlo what you think?

**Comment by roigcarlo** [comment]
If nobody is gonna use it ok, but I don't see any harm leaving them

**Comment by pooyan-dadvand** [comment]
More code, more maintenance...

**Comment by msandre** [comment]
I'm fine with either. It's a small amount of code but if one modifies one of them they have to remember to update the other or call it from the second version. Since it is enough to have one version we can put the reference version.

**Comment by pooyan-dadvand** [comment]
Actually if you look in the core you will find hundreds of methods with overrides to help in argument conversions. So it was my style too. But now I see that many of them have not used ever and they only give the headache that in type of tracing your problem you don't now which of them is used and which of them may be absolute by time and other changes. 

Of course this was my suggestion and I don't have real objection about it.

**Review by pooyan-dadvand — APPROVED** [review]

**Comment by roigcarlo** [comment]
Ok I will merge since this is quite important to fix this before 5.3. Nobody is using it except model_part_io and @msandre who is aware of the discussion, so in any case we can change it later on.

**Comment by loumalouomega** [comment]
Well @annasophie91 will use it

**Pull request merged**